megatron/core/optimizer/__init__.py

```diff
@@ -307,7 +277,14 @@ def _get_megatron_optimizer_based_on_param_groups(
     # when freezing sub-models we may have no trainable parameters on a rank and
     # hence an empty param_groups. However, we still need to create an optimizer
     # for the purposes of grad stats reductions
+    # TODO: All passed-in param_groups should share the same higher-level config object
+    # parameters (e.g., optimizer, optimizer_cpu_offload, etc.)
+    # TODO: logic needs to be updated to handle different optimizer types (i.e., param_groups
+    # passed into this function need to correspond to the same optimizer).
+
     if param_groups:
+        # TODO: Delete from param_groups object?
+        # TODO: What if optimizer is different for each config object?
         if config.optimizer_cpu_offload:
```

Comment:
Add an assertion for now to make sure the optimizer is the same for each overridden object.